[Merge #1] Gh 34, 49, 55, 60, 65: Add All Store Types & Schema To Create-Graph, Retrieve Graph URLs

macenturalxl1 opened the pull request:
This pull request has the changes to create graphs to have either a Accumulo or Federated Store.

This contains the original pull request changes for Accumulo Store: https://github.com/gchq/gaffer-experimental/pull/46 - as the Federated Store request needed some of these changes to make the request and can be reviewed here

Also the UI has been updated to send requests to create Accumulo or Federated Store Graphs

macenturalxl1 commented:
TODO (from gh-34 Accumulo Store PR review):
`gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/services/CreateGraphService.java`
"In the future you could convert to a Gaffer object as found in the gaffer-controller and move gaffer-controller the model to the gaas-common. This is a more generic option with a LinkedHashMap based GraphSpec rather than the more specialised CRDCreateRequestBody. At the end of the day both give the same JSON output "

macenturalxl1 commented:
Hi @r32575 , the API is now using `GafferSpec.java` to create a Gaffer Values Kubernetes object. Using a `GafferSpecBuilder.java` to build the spec in a more developer-friendly way. 

We can put it in to this branch and when this one is reviewed/approved first, it can be merged in to the folder structure branch for the other pull request

r32575 commented on gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/GafferSpecBuilder.java:
```diff
@@ -0,0 +1,123 @@
+            case ACCUMULO:
+                // No AccumuloStoreProperties required for the graph
+                // Instead, enableAccumulo() in the GraphSpec to enable Accumulo store properties
+                throw new IllegalArgumentException("enableAccumulo() in the GraphSpec to enable Accumulo store properties");
```
Rather than throwing an exception could you create an empty `stroreProperties` or leave it `null` and log a warning?
It's a bit messy having a `storeProperties` object and having a `enableAccumulo` method. It would be nice to have different objects for each store `GafferSpec`. That way you could get rid of the `enableAccumulo` and have the logic in an `AccumuloGraphStore` which extends `AbstractGraphSpec` (with the `graphId` and `description` logic in) and implements `GraphSpec`. You could then call the requested store type in the `GafferHelmValuesFactory.createGafferSpecFrom` method.

macenturalxl1 replied:
- I'll remove it and use the abstract builder solution. I added the IAX just for future devs working on it if they add implementation in such a way they'd get the illegal arg feedback from code to hint how the helm Values is configured using Accumulo - use the abstract builder does and reads the same way more cleanly so will go with that

- I've created an `AbstractGafferSpecBuilder` for a 'generic' `GafferSpecBuilder` (used for all other stores) and `AccumuloGafferSpecBuilder` to extend, is this what you mean? 

I've pushed the code for this. I've left the default storeproperties values added in the builder as I made the assumption Users of the API will just want to create a Graph with a generic Proxy Store for example, and might not what StoreProps to use - would this be correct? 
In another ticket, I can add custom storeProperties to the request if another (more technical) User Group knows how store props work and would add their own custom properties for their graph

r32575 commented on gaas-rest/pom.xml:
```diff
@@ -55,6 +56,11 @@
+        <dependency>
+            <groupId>uk.gov.gchq.gaffer</groupId>
+            <artifactId>federated-store</artifactId>
+            <version>1.13.3</version>
```
Version should be in properties. You seem to be using version `1.13.3` and `1.13.0` of `Gaffer`. It gets inherited by the parent pom so you may not even need versions. We have just moved to `1.16.0` so may want to use that.

macenturalxl1 replied:
Done

r32575 commented:
I was chatting to @GCHQDev404 about this. We were thinking that you could create separate objects that extend the GraphSpec such as AccumuloGraphSpec or FederatedGraphSpec. You could use Spring to create a list of the objects that use the GraphSpec interface. Each could return the type of graph they support such as accumulo or federated so when you have a graph type you could compare it (case insensitively) with the values given for each and if they match then you could get the builder from the GraphSpec from the matching object and build in much the same way. This would then mean you don't have to alter the code every time you add more GraphSpec objects.

This will probably be best to put in a separate task. I will leave a few requests on the code.

macenturalxl1 commented:
> I was chatting to @GCHQDev404 about this. We were thinking that you could create separate objects that extend the GraphSpec such as AccumuloGraphSpec or FederatedGraphSpec. You could use Spring to create a list of the objects that use the GraphSpec interface. Each could return the type of graph they support such as accumulo or federated so when you have a graph type you could compare it (case insensitively) with the values given for each and if they match then you could get the builder from the GraphSpec from the matching object and build in much the same way. This would then mean you don't have to alter the code every time you add more GraphSpec objects.
> 
> This will probably be best to put in a separate task. I will leave a few requests on the code.

Cool thank you for the feedback. It was something thought at first, and then considered the 'composition vs/over inheritence' so that's why we implemented the single GraphSpecBuilder and just add what ever overrides are required for the helmchart Values object for any incoming requests. 

Can put it in a new ticket - maybe add a label called recommendations for similar situations under consideration like this

macenturalxl1 commented:
Hi @r32575 I've just added the schema to the create-graph-request and retrieve URLs for each graph features to this PR - as it is some overlap with adding it to GraphSpec to send to the Controller

Just thought it saves having to look at multiple PRs that cover near enough the same thing

Also I've refactored an alternate solution to the one you mentioned in your earlier comment - let me know what you think

r32575 reviewed (changes requested):
I will raise a fresh ticket to have different GraphSpec objects to make `gaas-rest` maintainable in the future.

The project does not build at present with the following cause when running the controller tests:
```
Caused by: java.lang.NoSuchMethodError: com.google.gson.JsonParser.parseReader(Ljava/io/Reader;)Lcom/google/gson/JsonElement;
	at io.kubernetes.client.util.KubeConfig.runExec(KubeConfig.java:330) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.KubeConfig.tokenViaExecCredential(KubeConfig.java:269) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.KubeConfig.getAccessToken(KubeConfig.java:231) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.credentials.KubeconfigAuthentication.<init>(KubeconfigAuthentication.java:46) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.ClientBuilder.kubeconfig(ClientBuilder.java:283) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.ClientBuilder.getClientBuilder(ClientBuilder.java:115) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.ClientBuilder.standard(ClientBuilder.java:94) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.ClientBuilder.standard(ClientBuilder.java:86) ~[client-java-11.0.0.jar:na]
	at io.kubernetes.client.util.ClientBuilder.defaultClient(ClientBuilder.java:68) ~[client-java-11.0.0.jar:na]
	at uk.gov.gchq.gaffer.gaas.AppConfig.apiClient(AppConfig.java:32) ~[classes/:na]
	at sun.reflect.GeneratedMethodAccessor104.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_282]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.3.jar:5.3.3]
	... 99 common frames omitted
```
I don't know if it is just my VM that is playing up or not. I merged the folder structure branch so it may work when you update with the changes. I'll have another go when the branch has been updated.

r32575 commented:
Seems that there are still issues with running the `gaas-rest` tests with `kubectl` not pointing at a working cluster. I spun up `minikube` and the tests ran successfully. I stopped `minikube` and the tests failed.

r32575 commented:
Still finding that the tests fail if `kubectl` is pointing at non-existent cluster (or is turned off)... I would stop merging new branches in. If you want to work on a new branch that has all the changes from this branch in then create a new create a new branch and merge this branch into the new branch.

rocky341 commented:
@r32575 
Looking into the failing tests issue. If you do a "mvn clean install" on gaas-rest are there failing tests and if so which ones?

r32575 commented:
Hi @rocky341, just updated and ran tests again and it succeeded. No changes were made from your end so it was on my side so sorry and ignore the last message. I'll go through the changes today.

rocky341 commented on gaffer-as-a-service/gaas-rest/deploy/helm/values.yaml:
```diff
@@ -15,8 +15,8 @@
-  pullPolicy: IfNotPresent
+  repository: kitkat520/api
+  pullPolicy: Always
   tag: latest
```
Done

r32575 replied:
It is still saying `kitkat520/api` rather than `dockerhub/repo`

rocky341 replied:
It should be updated. Perhaps try to pull again?

rocky341 commented on gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/model/GaaSRestApiException.java:
```diff
@@ -14,7 +14,7 @@
  */
 
-package uk.gov.gchq.gaffer.gaas.exception;
+package uk.gov.gchq.gaffer.gaas.model;
```
Yes it should thank you. This is now done.

rocky341 commented on gaas-rest/deploy/helm/values.yaml:
```diff
@@ -15,8 +15,8 @@
-  repository: dockerhub/repo
-  pullPolicy: IfNotPresent
+  repository: kitkat520/api
+  pullPolicy: Always
```
Done
